extensions/amp-mega-menu/0.1/amp-mega-menu.js

```diff
@@ -94,79 +114,99 @@ export class AmpMegaMenu extends AMP.BaseElement {
     // in which case register them after DOM update instead.
-    this.element.addEventListener(AmpEvents.DOM_UPDATE, () => {
-      this.registerMenuItems_();
-    });
+    this.element.addEventListener(AmpEvents.DOM_UPDATE, this.domUpdateHandler_);
+
+    if (!this.maskElement_) {
+      const mask = this.document_.createElement('div');
+      mask.classList.add('i-amphtml-mega-menu-mask');
+      mask.setAttribute('aria-hidden', 'true');
+      // append mask to header so that all children of header appear above the mask
+      const maskParent =
+        closestAncestorElementBySelector(this.element, 'header') ||
+        this.element;
+      maskParent.classList.add('i-amphtml-mega-menu-mask-parent');
+      maskParent.appendChild(mask);
+      this.maskElement_ = mask;
+    }
 
-    const mask = this.document_.createElement('div');
-    mask.classList.add('i-amphtml-mega-menu-mask');
-    mask.setAttribute('aria-hidden', 'true');
-    // append mask to header so that all children of header appear above the mask
-    const maskParent =
-      closestAncestorElementBySelector(this.element, 'header') || this.element;
-    maskParent.classList.add('i-amphtml-mega-menu-mask-parent');
-    maskParent.appendChild(mask);
-    this.maskElement_ = mask;
-
-    this.documentElement_.addEventListener('click', () => this.collapse_());
-    this.documentElement_.addEventListener('keydown', event => {
-      // Collapse mega menu on ESC.
-      if (event.key === Keys.ESCAPE && this.collapse_()) {
-        event.preventDefault();
-      }
-    });
+    this.documentElement_.addEventListener('click', this.documentClickHandler_);
+    this.documentElement_.addEventListener(
+      'keydown',
+      this.documentKeyDownHandler_
+    );
 
     return Promise.resolve();
   }
 
   /** @override */
   unlayoutCallback() {
-    // TODO(#25047): unregister event listeners.
+    this.element.removeEventListener(
+      AmpEvents.DOM_UPDATE,
+      this.domUpdateHandler_
+    );
+    this.documentElement_.removeEventListener(
+      'click',
+      this.documentClickHandler_
+    );
+    this.documentElement_.removeEventListener(
+      'keydown',
+      this.documentKeyDownHandler_
+    );
+    this.unregisterMenuItemListeners_();
     // Ensure that menu is closed when hidden via media query.
     this.collapse_();
-    return false;
+    return true;
   }
 
   /**
-   * Find all expandable menu items under this mega menu and add appropriate
-   * classes, attributes and event listeners to its children.
+   * Find all menu items under this mega menu and add appropriate classes,
+   * attributes and event listeners to its children.
    * @private
    */
   registerMenuItems_() {
-    this.items_ = toArray(scopedQuerySelectorAll(this.element, 'nav > * > li'));
-    this.items_.forEach(item => {
+    // remove existing event listeners and re-add them at the end.
+    this.unregisterMenuItemListeners_();
```

Comment:
There is no need to do this. Adding the same listener twice (if it is the same function reference) does not cause duplicate listeners to be added.